### PR TITLE
Fair adult - WIP

### PR DESCRIPTION
**Pull request by ayushi-3536**
- Fair adult benchmark from [MOASHA paper](https://arxiv.org/abs/2106.12639)
- added adult dataset preprocessing pipeline 
-  added config space and objective function
- MOASHA records information every epoch - helpful for benchmarking / early-stopping
  ( not implemented here for now, need to decide parameters to register in that case)
- added dependencies
- added fairness metric - (code structuring is in progress but utility will remain same)
- Testing In Progress
- Containerization in Progress

### CODE REVIEW

File: hpobench/util/fairness_metrics.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+TPR1 = 'tpr1'
+TPR_DIF = 'tpr_dif'
+TPR_MIN = 'tpr_min'
+
+FAIRNESS_METRICS = [STATISTICAL_DISPARITY, UNEQUAL_OPPORTUNITY, UNEQUALIZED_ODDS, TPR0, TPR1, TPR_DIF, TPR_MIN]
+
+PRED_THRESHOLD = 0.5
+
+
+def fairness_risk(x, y, sensitive_rows, model, unfairness_metric):
```
Comment (pfistfl) [inline hpobench/util/fairness_metrics.py]: Are you sure you do not wanna just use the metrics from `fairlearn`? Or is that the original code?

Comment (ayushi-3536) [reply hpobench/util/fairness_metrics.py]: -- The metric calculation is taken from original code but this should be doable by fairlearn utility aswell
-- will integrate fairlean to keep it clean

Comment (pfistfl) [reply hpobench/util/fairness_metrics.py]: Hmmm… perhaps keep the original code, then we are closer to the original work

File: hpobench/benchmarks/mo/adult_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+        cs = CS.ConfigurationSpace(seed=seed)
+
+        cs.add_hyperparameters([
+            CS.UniformIntegerHyperparameter(
+                'n_fc_layers', default_value=3, lower=1, upper=4, log=False
+            ),
+            CS.UniformIntegerHyperparameter(
+                'fc_layer_0', default_value=16, lower=2, upper=32, log=True
+            ),
+            CS.UniformIntegerHyperparameter(
```
Comment (KEggensperger) [inline hpobench/benchmarks/mo/adult_benchmark.py]: Shouldn't these be conditional on n_fc_layers? The optimizer could still choose to ignore that, but if possible it can also use that additional info

File: hpobench/benchmarks/mo/adult_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+        start = time.time()
+        y_pred_test = mlp.predict(X_test)
+        test_accuracy = accuracy_score(self.y_test, y_pred_test)
+        eval_test_runtime = time.time() - start
+
+        val_statistical_disparity = fairness_risk(X_valid, self.y_valid, sensitive_rows_val, mlp, STATISTICAL_DISPARITY)
+        val_unequal_opportunity = fairness_risk(X_valid, self.y_valid, sensitive_rows_val, mlp, UNEQUAL_OPPORTUNITY)
+        val_unequalized_odds = fairness_risk(X_valid, self.y_valid, sensitive_rows_val, mlp, UNEQUALIZED_ODDS)
+
+        test_statistical_disparity = fairness_risk(X_test, self.y_test, sensitive_rows_test, mlp, STATISTICAL_DISPARITY)
```
Comment (KEggensperger) [inline hpobench/benchmarks/mo/adult_benchmark.py]: This costs extra time. Is there a reason to report test performance?

File: hpobench/util/data_manager.py [hunk 4]
```diff
@@ -926,6 +926,165 @@ def _load(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndar
         return X_trn, y_trn, X_val, y_val, X_tst, y_tst
 
 
+class AdultDataManager(HoldoutDataManager):
+
+    def __init__(self):
+        super(AdultDataManager, self).__init__()
+        self.logger.debug('AdultDataManager: Starting to load data')
+        self.urls = {"data": "http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",
```
Comment (KEggensperger) [inline hpobench/util/data_manager.py:934]: Is this a different dataset than the adult dataset on OpenML? https://www.openml.org/d/1590
If not, I would argue for using the OpenML dataset for this or is there a reason not to?

**Review by KEggensperger — APPROVED** [review]
Looks good to me, I just left a few comments

File: hpobench/container/benchmarks/mo/adult_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+from hpobench.container.client_abstract_benchmark import AbstractBenchmarkClient
+
+
+class AdultBenchmark(AbstractBenchmarkClient):
+    def __init__(self, **kwargs):
+        kwargs['benchmark_name'] = kwargs.get('benchmark_name', 'AdultBenchmark')
+        kwargs['container_name'] = kwargs.get('container_name', 'fair_adult')
+        kwargs['latest'] = kwargs.get('container_tag', '0.0.1')
+        kwargs['gpu'] = kwargs.get('gpu', True)
```
Comment (KEggensperger) [inline hpobench/container/benchmarks/mo/adult_benchmark.py]: Does this mean this benchmark requires a GPU? If yes, where?

File: hpobench/benchmarks/mo/adult_benchmark.py (hunk 3 above)
Comment (PhMueller) [reply hpobench/benchmarks/mo/adult_benchmark.py]: We thought that we can use it for a HV plot on the test set via time. But as discussed in the last meeting, this might be not necessary now. 

@noorawad: do you have an opinion on that?

File: hpobench/util/data_manager.py (hunk 4 above)
Comment (PhMueller) [reply hpobench/util/data_manager.py:934]: @ayushi-3536 Can you answer this on?

File: hpobench/util/fairness_metrics.py (hunk 1 above)
Comment (KEggensperger) [reply hpobench/util/fairness_metrics.py]: If there is original code, would it be possible to import this code? Or at least leave a comment saying where the original code is?

File: hpobench/util/data_manager.py (hunk 4 above)
Comment (ayushi-3536) [reply hpobench/util/data_manager.py:934]: Both the datasets have same sources([UCI](https://archive.ics.uci.edu/ml/datasets/Adult) - 1996), However the one used already process a few fields and split the data into test and training sets. More info here([http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.names). Since we are using code used in the paper for preprocessing I thought it would be safer to stick with the processed data they used.
One instance of how it is related :
-- The "Unknown" values in openml dataset are replaced with '?' in the datasets used and  in data processing code '?' replaced with nan.

File: hpobench/benchmarks/mo/adult_benchmark.py (hunk 3 above)
Comment (PhMueller) [reply hpobench/benchmarks/mo/adult_benchmark.py]: For the maximal network size (4 layers à 32 units), the evaluation per split takes around 0.05 to 0.1 seconds.
I would suggest to include the metrics for all splits, since it is not that expensive. 

And it might be interesting for later evaluations.
